### PR TITLE
Make find(list) not auto-null on a behavior of 2.x

### PR DESCRIPTION
**Pull request by dereuromark**
... where select(keyfield, valuefield) would autopopulate.

Makes this work again:

```
$query = $table->find('list', ['fields' => ['id', 'created']]);
```

or

```
    $query = $table->find('list', ['fields' => ['id']]);
```

Resolves https://github.com/cakephp/cakephp/issues/8097

For all cases where keyField or valueField are still not part of the select clause we might still want to throw an exception maybe.

### CODE REVIEW

File: src/ORM/Table.php [hunk 1]
```diff
@@ -1011,6 +1011,14 @@ public function findAll(Query $query, array $options)
      */
     public function findList(Query $query, array $options)
     {
+        if (!isset($options['keyField']) && !isset($options['valueField'])) {
+            $select = $query->clause('select');
+            if ($select && count($select) <= 2) {
+                $options['keyField'] = array_shift($select);
+                $options['valueField'] = array_shift($select) ?: $options['keyField'];
```
Comment (lorenzo) [inline src/ORM/Table.php]: What happens when the field is in he form of "table.field"? What if the user tries to select a field from another table?

Comment (dereuromark) [inline src/ORM/Table.php]: In any of those cases the select has been issued, so if it fails, then because he did not also join (contain) that association.
For all other cases one **must** manually set keyField and valueField of course.
This here only covers the trivial case of having two simple selects but no keyField/valueField selected AND otherwise the *wrong* fields selected silently.

Comment (lorenzo) [inline src/ORM/Table.php]: the query has not been executed at this point, but that is actually not the problem here. What I meant is that the fields syntax as accepted by select() is different from what keyField and valueField accepts.

If you want to try it yourself, add a test where you use 'Articles.id' as one of the fields in select()

Comment (dereuromark) [inline src/ORM/Table.php]: OK, in those cases one has to add the key and value field definitions.
But run my tests without my patch and you will see that it uses two totally different fields and resulting in null rows without the expectation of it.
If you later at least would run into some error it would be clear what is going on.
But currently fields are silently changed without the intention to do so.

Comment (lorenzo) [inline src/ORM/Table.php]: So using Articles.id as a field yielding null values is ok to you in this case?

Comment (dereuromark) [inline src/ORM/Table.php]: I don't get your point.
Currently it does not work at all - this fixes it.
Current behavior using the documented definitions is not affected.
So it only makes things better.

Do you have a test case that shows what you mean?

Comment (lorenzo) [inline src/ORM/Table.php]: Just add a test where you use Articles.id instead of id and you will see what I mean.

Comment (dereuromark) [inline src/ORM/Table.php]: It throws a `Undefined property: Cake\Test\TestCase\ORM\TableTest::$Users` notice, which at least tells you something, that this aint working. You will then at some point at least realize that you need to add keyField/valueField.
Or we could extract the model alias here and throw it away for the definition setting.
If that makes more sense.

**Comment by ADmad** [comment]
I don't really like this change. I would prefer if we don't introduce old automagic and stick with having to explicitly specify `keyField` and `valueField`.

Since the change is primarily meant to make upgrading easier, those who require it can override the method and add those extra lines. This belongs in a shim plugin IMO.

**Comment by ionas** [comment]
:+1: @ADmad
The only thing I'd make sure is that in case find('list') does not return any fields and/or virtual properties that are specified in keyField/valueField during runtime, it should throw an exception so that there is no sudden surprise that some overlooked select is not filled up on the FormHelper.

**Comment by dereuromark** [comment]
@ADmad We already have (bad) automagic that breaks code in an unexpected way.
Please see @markstory last comment on the original issue. I think we can agree on a more sane fallback here for those cases where the currently included automagic does the wrong thing.
And as discussed we cannot that easily throw exceptions here due to a few reasons.

**Comment by ionas** [comment]
I'd rather see the exceptions being fixed at a later stage in code than having more magic that will possibly confuse *future* users.

*Why would it not work to the moment `find('list')` should return keyField and valueField hash (either set explicitly by the user or implicitly to defaults) if there are neither "real" properties nor "virtual" properties matching both, keyField and valueField?*

**Comment by dereuromark** [comment]
It is not more magic, it is less wrong magic (if you select manually you would expect those to be taken into account instead of magically setting two wrong fields that can't possibly be part of the resultset).

For the record: I am fine for this being part of Shim plugin, as this is where I put it [for the moment](https://github.com/dereuromark/cakephp-shim/commit/ff89bf1251ce1207ffb7e78c4d8f7ee4cfa676da).
But I would highly recommend this bugfix to be applied to the core for all people tripping over the current bad magic behavior.

**Comment by dereuromark** [comment]
So what is the course of action here?
While I still think this is mainly interesting for the ones upgrading, I also think that if you provide 2 select fields manually you would want those to end up being used instead of 2 totally wrong ones selected.
If we can get the cases to work that have been mentioned above (Table prefixing), I think this would be a good fallback addition either way. But even without the prefixing working you at least get a better development error now (missing fields) instead of silently chosing the wrong ones and ending up with empty form selects.
Can someone cook up a test case that is not yet covered?

**Comment by markstory** [comment]
Given that this appears to handle table.field select values it seems more reasonable. However, I agree with thr others that this implicit behavioyis not ideal long term. Would it be unreasonable to trigger a deprecated warning ehen this situation occurs? That way we could do what the user might have wanted and let them know about how it should be done now at the same time?

**Comment by dereuromark** [comment]
As long as the other implicit behavior (auto select some key and value field) is still in place, this auto select of select query parts is as valid IMO.
Or differently: The auto field selection with any custom query select is quite dangerous without checking for those fields to be present.

So IMO we either remove both or keep this as a smart way of selecting the right fields.
I would be fine with both, but prefer the latter (it hurts no one).

**Comment by dereuromark** [comment]
I added a few more test cases for those edge cases you meant.
But I am not sure if we should auto-join the relations, it would be better to let the ORM throw the missing contain exception IMO.

**Comment by dereuromark** [comment]
Should be solved now.

**Comment by markstory** [comment]
Looks good to me.

**Comment by lorenzo** [comment]
Good job @dereuromark

**Comment by jippi** [comment]
This caused a regression in all bownty apps :) 

````
		$countries = TableRegistry::get('Locations')->find('list', [
			'fields' => ['id', 'name' => 'name_english'],
			'conditions' => ['is_active' => 1, 'parent_id is null'],
			'order' => ['name_english']
		])->all();
````

no longer works - the `name` (value) field is now always `NULL` cc @lorenzo

**Comment by lorenzo** [comment]
let the bugs commence! :P

**Comment by dereuromark** [comment]
How could this have worked without specifying key and value fields? Is that from the translated behavior?

**Comment by lorenzo** [comment]
@dereuromark `name` is the displayField, so it was picked up automatically

**Comment by ADmad** [comment]
Automagic is the path to failure!

**Comment by jippi** [comment]
I'm voting for reverting this magic and just call it an "oopsie" :)

**Comment by dereuromark** [comment]
I vote to completely remove the automagic as it should have been and call it a day.

**Comment by markstory** [comment]
@dereuromark Removing the old behavior would break applications as well.

**Comment by ionas** [comment]
Removing in 2.x or 3.x?

**Comment by jippi** [comment]
3.x only
